Add new image container component and deprecate legacy image variants

jmuzina opened the pull request:
## Done

Added several classes that will help unify the appearance of images in projects that use Vanilla, and deprecated some styles that are no longer part of branding.

- `p-image-container--` classes for setting element aspect ratio.
    - `p-image-container--16-9`
    - `p-image-container--3-2`
    - `p-image-container--2-3`
    - `p-image-container--cinematic`
    - `p-image-container--square`
- `p-image-container .is-highlighted` for distinguishing an image from its surroundings by applying a slight underlay.
- Deprecated image modifiers `p-image--bordered` and `p-image--shadowed`

Closes: #4876 [WD-11412](https://warthogs.atlassian.net/browse/WD-11412)

## QA

- Review updated [Images pattern documentation](https://vanilla-framework-5097.demos.haus/docs/patterns/images)
- Review [updated release notes for 4.12.0](https://vanilla-framework-5097.demos.haus/docs/whats-new)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [x] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix relesase (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).


## Screenshots

Background highlight:
![image](https://github.com/canonical/vanilla-framework/assets/46915153/61aef1ea-0d16-4665-9edc-30f1fe4dce87)
Aspect ratio:
![image](https://github.com/canonical/vanilla-framework/assets/46915153/bf270843-6437-4d80-b9c5-407d2ecc8e5f)


[WD-11412]: https://warthogs.atlassian.net/browse/WD-11412?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://vanilla-framework-5097.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/13701)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

lyubomir-popov commented:
hi - I think we can drop `u-image-container--padded` as I made an issue to keep hr margins in all cases, so we no longer need to do this. And I think the name of 2-3-4 should be --2.4 (was using 2.34 but the correct one is 2.4 after reading [this ](https://www.panamorph.com/which-home-theater-screen-aspect-ratio-is-best-2-401-2-351-or-169/#:~:text=2.35%3A1%20is%20the%20aspect,movie%20aspect%20ratios%20out%20there))

jmuzina commented:
As @lyubomir-popov mentioned, `u-image-container--padded` can be removed as we will be deprecating `p-rule`. See #5096

jmuzina commented:
> I think the name of 2-3-4 should be --2.4 (was using 2.34 but the correct one is 2.4 after reading [this ](https://www.panamorph.com/which-home-theater-screen-aspect-ratio-is-best-2-401-2-351-or-169/#:~:text=2.35%3A1%20is%20the%20aspect,movie%20aspect%20ratios%20out%20there))

Adjusted this to use the class name you described and switched to `aspect-ratio: 1/2.4`.

While it is syntactically possible for us to use `u-aspect-ratio--2.34`, the period in the class name has to be [escaped](https://github.com/jmuzina/vanilla-framework/blob/ff20ec6b7e2d2c40d1a6edbf3c3c71c396ffd4ea/scss/_utilities_aspect-ratio.scss#L11) for the CSS to be valid (`u-aspect-ratio--2\.34`). Then it can be [used as normal](https://github.com/jmuzina/vanilla-framework/blob/ff20ec6b7e2d2c40d1a6edbf3c3c71c396ffd4ea/templates/docs/examples/utilities/aspect-ratio/1-2-4.html#L12).

This causes a [linting failure](https://github.com/canonical/vanilla-framework/actions/runs/9211205543/job/25340035937?pr=5097) due to our [linter configuration](https://github.com/canonical/vanilla-framework/blob/d0197b27f5733789d6d23536eabe68356e045ae1/.stylelintrc.json#L11).

@bartaz @pastelcyborg What do you think is the most semantically clear way to name this class? I don't think we should alter our linting rules for this. Perhaps `u-aspect-ratio--theater` as suggested by the link @lyubomir-popov sent?

bartaz commented:
I still think we should not introduce them as utilities if they are meant for images only.

Utilities should be a last resort class name that clearly can be applied in ANY context. These from their description seem like are designed to be used with images and images only.

Our current image styling is quite dated (we probably don't want to use the `p-image--bordered` or `p-image--shadowed` anymore https://vanillaframework.io/docs/patterns/images), but maybe it's an opportunity to clean that up and provide a consistent image component with options that we actually need and use.

Creating those image-only class names as utilities is a bad precence in my opition.

@lyubomir-popov What kind of images styles do we need to support?

I see we have images that cover all available space, and illustrations that have some kind of highlight background around them. Any other "background" styles?

The other variable is aspect ratio that is quite well defined.

Is there anything else about images we want to control?

How I see it is that we could possibly have some kind of `p-image` component with any options needed.

So something like:

```<figure class="p-image--aspect-ratio-3-4 is-highighted><img class="p-image__img" /></figure>```

pastelcyborg commented:
Thanks for making these changes. I'm wondering if we should change the examples to make it a bit easier to tell what each of them is actually doing; because the image used is so small and there's no background color or border, it's kind of difficult to understand the aspect ratio being applied, since the examples all kind of just look like a big white frame. Even just making the example embed smaller (particularly horizontally) might help here.

pastelcyborg commented on scss/_patterns_image.scss:
```diff
@@ -1,6 +1,37 @@
+    }
+  }
+
+  [class^='p-image-container'] {
```
I'd probably make this:
```
.p-image-container,
[class^='p-image-container--'] {
```
Because using the starts-with selector on the entire Block class means it would apply to any Elements also, if we ever were to add any, and that's probably not desirable.

pastelcyborg commented on scss/_patterns_image.scss:
```diff
@@ -1,6 +1,37 @@
+    aspect-ratio: 1/1;
+  }
+
+  .p-image-container--highlighted-background {
```
This also kinda throws a wrench in the starts-with selector above, since it would create a selector containing duplicate styles from the `common` placeholder.

bartaz commented on scss/_patterns_image.scss:
```diff
@@ -1,6 +1,38 @@
+  }
+
+  [class^='p-image-container'].is-highlighted {
+    background: rgba(0, 0, 0, 0.05);
```
This would need to come from variable, possibly themed one. So we beed a value for both light and dark theme.

@lyubomir-popov What would be the dark version?

lyubomir-popov replied:
pls use the variable called background-alt

bartaz commented on scss/_patterns_image.scss:
```diff
@@ -1,6 +1,38 @@
+    @extend %vf-image-container--common;
+  }
+
+  [class^='p-image-container'].is-highlighted {
```
Ideally this would be the same selector as the one above in line 10.

I guess for simplicity we can just nest `&.is-highlighted` inside there?

jmuzina replied:
Moved this (& the centering CSS to remove the alignment utilities from the example) to the `vf-image-container--common` placeholder for simplicity - is that suitable?

bartaz commented on templates/docs/examples/patterns/image/container/aspect-ratio/16-9.html:
```diff
@@ -0,0 +1,10 @@
+{% block standalone_css %}patterns_image{% endblock %}
+
+{% block content %}
+<div class="p-image-container--16-9 u-align--center u-vertically-center">
```
Utilities shouldn't be required part of a component. They should only be needed if you need to modify the default behaviour that was not specified by the component itself.

If the image is supposed to be horizontally and vertically centered in the container by default, that's what the `p-image-container--` should do itself ideally.

jmuzina replied:
Revised (see [this comment](https://github.com/canonical/vanilla-framework/pull/5097#discussion_r1618861457)) and added note to documentation about centering behavior & how to change image position.

bartaz commented:
@jmuzina @pastelcyborg What do you think about https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit https://developer.mozilla.org/en-US/docs/Web/CSS/object-position

At the first glance it seems like it could help achieve what we want with just an `img` element. Because as far as I can see we could even set the CSS background on img itself.

bartaz commented on templates/docs/examples/patterns/image/container/aspect-ratio/16-9.html:
```diff
@@ -0,0 +1,10 @@
+
+{% block content %}
+<div class="p-image-container--16-9 u-align--center u-vertically-center">
+  <img src="https://res.cloudinary.com/canonical/image/fetch/f_auto,q_auto,fl_sanitize,c_fill,w_720/https://assets.ubuntu.com/v1/0ecd2ffd-community_wallpaper_noble_little_numbat_boy_by_azskalt.jpeg" alt="">
```
I think it's OK to use direct `https://assets.ubuntu.com/v1/...` links in Vanilla examples. Cloudinary links are usually generated by our image template, so I'd rather not hardcode them into examples.
And it will be also good to avoid using third party service to serve these images as this can impact on Percy snapshots being less stable.

jmuzina replied:
Revised to use assets API directly

bartaz commented on templates/docs/examples/patterns/image/container/aspect-ratio/2-3.html:
```diff
@@ -0,0 +1,10 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Image Container / Aspect Ratio / 2/3{% endblock %}
```
I wonder if we need a separate example for every single aspect ratio.

There is probably little value for it, as we likely don't want to embedd them separately in the docs. They are quite self explanatory, so having one of them as an example, with a listing of other available class names should be enough.

Obviously for the sake of reference and testing we should have a single "Image container / All aspect ratios" example that would just showcase all of them all together?

jmuzina replied:
Revised to use two examples: [16:9](https://vanilla-framework-5097.demos.haus/docs/examples/patterns/image/container/aspect-ratio/16-9?theme=light) (embedded in [documentation](https://vanilla-framework-5097.demos.haus/docs/patterns/images#image-container-with-aspect-ratio) as this is the most frequent use case) and [all](https://vanilla-framework-5097.demos.haus/docs/examples/patterns/image/container/aspect-ratio/all?theme=light)

jmuzina commented:
> @jmuzina @pastelcyborg What do you think about https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit https://developer.mozilla.org/en-US/docs/Web/CSS/object-position

I've revised to use `object-fit: contain` as it seems to be more appropriate for our use case.

> At the first glance it seems like it could help achieve what we want with just an `img` element. 

Are you suggesting to get rid of the wrapping div with `p-image-container--` classes and switch to `p-image--`? I think we need the div with its own aspect ratio set so the image has a parent container with a set aspect ratio that it can be neatly contained within.

> Because as far as I can see we could even set the CSS background on img itself.

True, we can set the background on img itself. However if we do this and combine usage with the aspect ratio classes, the background is applied only to the image, not the entire container. You'd get a result like this:

![image](https://github.com/canonical/vanilla-framework/assets/46915153/9552b557-04d2-4a4b-9497-f9ff39061e2e)

Instead of the desired:
![image](https://github.com/canonical/vanilla-framework/assets/46915153/d4420dfc-11c1-4f3e-b1d1-553d953d42c1)

(images from [c.com mongodb page](https://canonical.com/data/mongodb))

bartaz commented:
> True, we can set the background on img itself. However if we do this and combine usage with the aspect ratio classes, the background is applied only to the image, not the entire container. You'd get a result like this:
> 
> (images from [c.com mongodb page](https://canonical.com/data/mongodb))

Thanks for exploration. Is the issue caused by the fact that our img are inline by default? Seems that setting them to block and 100% width (as they would be a div) seems to solve the issue.
I guess one downside I can see is that it may restrict our ability of setting width/height of actual image (because it would be controlled by object-fit and not width/height of the img element.

jmuzina commented:
> any idea when is this going to be released?

It could be early next week when @bartaz returns from holiday.

jmuzina commented:
> One downside I can see is that it may restrict our ability of setting width/height of actual image (because it would be controlled by object-fit and not width/height of the img element.

I think we should avoid doing it this way for this reason. I'd rather not restrict ability to set width & height, which the projects that use Vanilla do quite often.

See c.com: https://github.com/search?q=repo%3Acanonical%2Fcanonical.com%20width%3D&type=code

lyubomir-popov commented:
to summarise, is the discussion whether to put the background colour on a parent? If so, I think that's the right way to do it. The background fade sets the aspect ratio, and the image is just centered within.

jmuzina commented:
@lyubomir-popov 
> to summarise, is the discussion whether to put the background colour on a parent?

I believe Bartek is proposing that we implement all of this functionality (aspect ratio & background highlight) on just the image with no parent, at the cost of control over image width/height

pastelcyborg commented:
I would concur that there should be a container element used here. Removing that and having solely the <img> itself eliminates the ability of a lot of fine-grained manipulation, particularly to the image's spacing/sizing. It would also make debugging kind of strange, as the <img> would have a container and metrics totally different from the actual image itself. I have a gut feeling there are other implications to using the <img> in this way as well, but I'm having trouble articulating exactly what those are without further experimentation.

bartaz commented on templates/docs/examples/patterns/image/container/highlighted-background.html:
```diff
@@ -0,0 +1,10 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Image container / highlighted background{% endblock %}
```
```suggestion
{% block title %}Image container / Highlighted{% endblock %}
```

bartaz commented on templates/docs/patterns/images.md:
```diff
@@ -4,10 +4,46 @@ context:
-Enhance images by adding a variant style with a border or drop shadow.
+Enhance images by distinguishing them from their background or controlling their container's aspect ratio.
+
+## Image container with highlighted background
```
Maybe simply "Highlighted image"?
```suggestion
## Highlighted image container
```

jmuzina replied:
Revised to highlighted image container

bartaz commented on templates/docs/patterns/images.md:
```diff
@@ -4,10 +4,46 @@ context:
+<div class="p-notification--caution">
+  <div class="p-notification__content">
+    <h3 class="p-notification__title">Deprecated</h3>
+    <p class="p-notification__message">Image with border is deprecated. Use `.p-image-container--highlighted-background` instead.</p>
```
```suggestion
    <p class="p-notification__message">Image with border is deprecated. Use image container with `.is-highlighted` modifier instead.</p>
```

jmuzina replied:
revised

bartaz commented on templates/docs/patterns/images.md:
```diff
@@ -16,6 +52,13 @@ View example of image with border
+<div class="p-notification--caution">
+  <div class="p-notification__content">
+    <h3 class="p-notification__title">Deprecated</h3>
+    <p class="p-notification__message">Image with drop shadow is deprecated. Use `.p-image-container--highlighted-background` instead.</p>
```
```suggestion
    <p class="p-notification__message">Image with drop shadow is deprecated. Use image container with `.is-highlighted` modifier instead.</p>
```

jmuzina replied:
revised

bartaz reviewed (changes requested):
Several places still mention previous `--highlighted-background` instead of new `is-highlighted`. I'm not sure if I found all.

Probably the file `highlighted-background.html` could be renamed to `highlighted.html`, and there is mention in PR description.

bartaz commented:
@jmuzina @pastelcyborg OK. I agree that settling down on having a wrapping element is a safer approach, let's stick to that.

Now, how about the name of the component itself. Does it have to be "image container"? Or would "image" be enough, `p-image` `p-image--16-9`, etc? Assuming of course that this class name is placed on the wrapping element?

```html
<figure class="p-image--16-9 is-highlighted">
    <img src="…" alt="…" />
</figure>
```

Also, if we have the separate image element inside, and we want to control it we probably should add a class name to it as well? `p-image__img`? Or `p-image-container__image` (in current naming)? What do you think?

jmuzina commented:
@bartaz 

> Does it have to be "image container"? Or would "image" be enough, `p-image` `p-image--16-9`, etc? Assuming of course that this class name is placed on the wrapping element?

While @pastelcyborg  and I think  "image container" itself is a strange name, it's clearer and safer to use "image container" than to use "image", as the image container's styles are intended for usage on an element that wraps an image. If we use `p-image`, people may get the wrong idea and apply, say, `p-image--16-9` directly to an image, which is unintended usage.

> Also, if we have the separate image element inside, and we want to control it we probably should add a class name to it as well? `p-image__img`? Or `p-image-container__image` (in current naming)? What do you think?

I would normally agree, but I think this sets a different precedent from the media container, which just uses `<img>` tag:
https://github.com/canonical/vanilla-framework/blob/426d15144f7efde7c0dcf0ee52c73b26a1aacdb8/scss/_patterns_media-container.scss#L3-L8

We could make this use `p-image-container__image` as long as we later commit to changing the media container to use an element class name `p-media-container__image` as well.

jmuzina commented:
@bartaz , @pastelcyborg and I met and decided to use `.p-image-container__image` instead of `<img>` tag here. This way the pattern is stricter & also supports other image tags like `<figure>` and `<object>`.

bartaz commented on scss/_patterns_image.scss:
```diff
@@ -1,6 +1,41 @@
+
+  .p-image-container,
+  [class^='p-image-container--'] {
+    @extend %vf-image-container--common;
```
Is this placeholder only extended here? Maybe the styles can be defined here directly, as there doesn't seem to be a need for the abstraction yet.

jmuzina replied:
Yes, I abstracted it early so it would be easier to extend later; but I've gone ahead and moved its styling to this selector

bartaz commented on templates/docs/examples/patterns/image/container/highlighted.html:
```diff
@@ -0,0 +1,10 @@
+{% block standalone_css %}patterns_image{% endblock %}
+
+{% block content %}
+<div class="p-image-container is-highlighted u-align--center">
```
Is the `u-align--center` needed? This should be built into the component ideally.

jmuzina replied:
Yes I built this into the common styling and forgot to remove it from the highlighted example as I did with the aspect ratio examples. Thanks; removed!

bartaz reviewed (changes requested):
Couple of small comments on the code. It also seems that the branch needs to be rebased due to version changes in package.json

jmuzina commented on scss/_patterns_image.scss:
```diff
@@ -1,6 +1,41 @@
+    &.is-highlighted {
+      background: $colors--theme--background-alt;
+    }
+    align-items: center;
```
Revised

bartaz commented:
I feel that currently this example is misleading:
<img width="1445" alt="image" src="https://github.com/canonical/vanilla-framework/assets/83575/e5067319-1468-48e2-b2b2-d9e3c0583d60">

Because you don't really see the container that has the aspect ratio set, so it's not a good use case for this feature.

I think it makes more sense to use the example with `is-highlighted` background, so that you actually see the container having aspect ratio, with image centered within.
So I guess we could reuse the example from the `is-highlighted`, just add aspect ratio to it.

jmuzina commented:
> I feel that currently this example is misleading: <img alt="image" width="1445" src="https://private-user-images.githubusercontent.com/83575/339343051-e5067319-1468-48e2-b2b2-d9e3c0583d60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyODM4ODUsIm5iZiI6MTcxODI4MzU4NSwicGF0aCI6Ii84MzU3NS8zMzkzNDMwNTEtZTUwNjczMTktMTQ2OC00OGUyLWIyYjItZDllM2MwNTgzZDYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDEyNTk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlOTZmNDAzMTVjZDQ0OTc3ODVkMTM0ZGZkMzc5OTg1N2U1ZTczNmIzMGEyMDFkMTc2ODc5YjU4OWJlMDlhOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DuvDXzp3spX2twg6NDKdV4Kb4_N3hGr7umA-hx7jjAY">
> 
> Because you don't really see the container that has the aspect ratio set, so it's not a good use case for this feature.
> 
> I think it makes more sense to use the example with `is-highlighted` background, so that you actually see the container having aspect ratio, with image centered within. So I guess we could reuse the example from the `is-highlighted`, just add aspect ratio to it.



> I feel that currently this example is misleading: <img alt="image" width="1445" src="https://private-user-images.githubusercontent.com/83575/339343051-e5067319-1468-48e2-b2b2-d9e3c0583d60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyODM4ODUsIm5iZiI6MTcxODI4MzU4NSwicGF0aCI6Ii84MzU3NS8zMzkzNDMwNTEtZTUwNjczMTktMTQ2OC00OGUyLWIyYjItZDllM2MwNTgzZDYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDEyNTk0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlOTZmNDAzMTVjZDQ0OTc3ODVkMTM0ZGZkMzc5OTg1N2U1ZTczNmIzMGEyMDFkMTc2ODc5YjU4OWJlMDlhOTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DuvDXzp3spX2twg6NDKdV4Kb4_N3hGr7umA-hx7jjAY">
> 
> Because you don't really see the container that has the aspect ratio set, so it's not a good use case for this feature.
> 
> I think it makes more sense to use the example with `is-highlighted` background, so that you actually see the container having aspect ratio, with image centered within. So I guess we could reuse the example from the `is-highlighted`, just add aspect ratio to it.

Revised accordingly:
<img width="812" alt="Screenshot 2024-06-13 at 9 10 55 AM" src="https://github.com/canonical/vanilla-framework/assets/46915153/616c9870-349b-4c49-898a-12bbb5df590b">

bartaz commented on releases.yml:
```diff
@@ -1,3 +1,13 @@
+- version: 4.13.0
+  features:
+    - component: Images
```
```suggestion
    - component: Image container
```

bartaz commented on releases.yml:
```diff
@@ -1,3 +1,13 @@
+  features:
+    - component: Images
+      url: /docs/patterns/images
+      status: Updated
```
```suggestion
      status: New
```

bartaz commented on releases.yml:
```diff
@@ -1,3 +1,13 @@
+      url: /docs/patterns/images
+      status: Updated
+      notes: "We've added new classes for controlling image aspect ratio and background."
+    - component: Images
```
```suggestion
    - component: Images / Legacy variants
```

bartaz commented on releases.yml:
```diff
@@ -1,3 +1,13 @@
+    - component: Images
+      url: /docs/patterns/images
+      status: Updated
+      notes: "We've added new classes for controlling image aspect ratio and background."
```
```suggestion
      notes: "We've added a new image container component for controlling image aspect ratio and background."
```

bartaz reviewed (changes requested):
Small tweaks to release notes

bartaz commented on templates/docs/examples/patterns/image/container/highlighted.html:
```diff
@@ -0,0 +1,10 @@
+{% extends "_layouts/examples.html" %}
+{% block title %}Image container / Highlighted{% endblock %}
```
A bit of inconsistent naming:

```suggestion
{% block title %}Image Container / Highlighted{% endblock %}
```

bartaz commented on templates/docs/patterns/images.md:
```diff
@@ -4,10 +4,46 @@ context:
+All image containers center the `.p-image-container__image` element inside them by default.
+If you need to change image alignment within the image container, use the [image position utility](/docs/utilities/image-position).
+
+| CSS Class                      | Aspect ratio |
```
That reminded me that we actually have a feature to generate list of component class names from the code comment, like this: https://vanillaframework.io/docs/base/code#class-reference

Maybe it would be relatively simple to add something like that to image container? It's basically a yaml at the beginning of the scss file:
https://github.com/canonical/vanilla-framework/blob/main/scss/_patterns_code-snippet.scss#L1-L45

jmuzina replied:
Revised to use class reference for classname documentation

bartaz reviewed (approved):
Thanks for all the updates and discussions, all looks good!
